packages/@lwc/template-compiler/src/codegen/static-element.ts

```diff
@@ -0,0 +1,217 @@
+    } else {
+        let hasDynamicText = false;
+        // it is ElseBlock | ForBlock | If | BaseElement
+        node.children.forEach((childNode) => {
+            collectStaticNodes(childNode, staticNodes, state);
+
+            childrenAreStaticSafe &&= staticNodes.has(childNode);
+            // Collect nodes that have dynamic text ahead of time.
+            // We only need to know if the direct child has dynamic text.
+            hasDynamicText ||= isText(childNode) && !isStringLiteral(childNode.value);
+        });
+
+        // for IfBlock and ElseifBlock, traverse down the else branch
+        if (isConditionalParentBlock(node) && node.else) {
+            collectStaticNodes(node.else, staticNodes, state);
+        }
+
+        nodeIsStaticSafe =
+            isBaseElement(node) &&
+            !isCustomRendererHookRequired(node, state) &&
+            isStaticNode(node, state.config.apiVersion);
+
+        if (nodeIsStaticSafe && hasDynamicText) {
+            // Track when the static element contains dynamic text.
+            // This will alter the way the children need to be traversed to apply static parts.
+            // See transformStaticChildren below.
+            STATIC_ELEMENT_WITH_DYNAMIC_TEXT_SET.add(node as StaticElement);
+        }
+    }
+
+    if (nodeIsStaticSafe && childrenAreStaticSafe) {
+        staticNodes.add(node);
+    }
+}
+
+export function getStaticNodes(root: Root, state: State): Set<ChildNode> {
+    const staticNodes = new Set<ChildNode>();
+
+    root.children.forEach((childNode) => {
+        collectStaticNodes(childNode, staticNodes, state);
+    });
+
+    return staticNodes;
+}
+
+// The purpose of this function is to concatenate the contiguous text nodes into a single array
+// to simplify the traversing logic when generating static parts and serializing the element.
+export function transformStaticChildren(elm: StaticElement) {
+    const children = elm.children;
+    if (!children.length || !STATIC_ELEMENT_WITH_DYNAMIC_TEXT_SET.has(elm)) {
+        // The element either has no children or its children does not contain dynamic text.
+        return children;
+    }
+
+    if (STATIC_ELEMENT_TO_DYNAMIC_TEXT_CHILDREN_CACHE.has(elm)) {
+        return STATIC_ELEMENT_TO_DYNAMIC_TEXT_CHILDREN_CACHE.get(elm)!;
+    }
+
+    const result: (StaticChildNode | Text[])[] = [];
+    const len = children.length;
+
+    let current: StaticChildNode;
+    let next: StaticChildNode;
+    let contiguousTextNodes: Text[] | null = null;
+
+    for (let i = 0; i < len; i++) {
+        current = children[i];
+        if (!isText(current)) {
+            contiguousTextNodes = null;
+            result.push(current);
+        } else {
+            if (!isNull(contiguousTextNodes)) {
+                // Already in a contiguous text node chain
+                // All contiguous nodes represent an expression in the source, it's guaranteed by the parser.
+                contiguousTextNodes.push(current);
```

Comment:
If this is the last item in the list, we will never call `result.push(contiguousTextNodes ?? current);`. This seems like a bug.

Reply:
If it's the last item in the list and it reaches this block, it doesn't need to call `result.push(contiguousTextNodes ?? current);` because the `contiguousTextNodes` is already in the `result` array.

The idea behind this function is to keep all static (single text literal not in a set of contiguous text nodes) as text nodes not consolidated into an array.

I have some tests for the texts in the template compiler test fixtures, [this one ](https://github.com/salesforce/lwc/pull/4099/files#diff-3b9cc3a4118c48f419de50410efd358af63234be8d8003be67ae131e12eaf981)might help answer any questions.

Reply:
Also updated the comment to help clarify.